lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,402 @@
+
+        if orography_cube.units != 'm':
+            msg = "Orography cube units are {}, must be metres!"
+            raise ValueError(msg.format(orography_cube.units))
+
+        enforce_float32_precision([temperature_cube])
+
+        # Extract x/y co-ordinates.
+        x_coord = temperature_cube.coord(axis='x').name()
+        y_coord = temperature_cube.coord(axis='y').name()
+
+        # Extract orography and land/sea mask data.
+        orography_data = next(orography_cube.slices([y_coord,
+                                                     x_coord])).data
+        land_sea_mask = next(land_sea_mask_cube.slices([y_coord,
+                                                        x_coord])).data
+        # Fill sea points with NaN values.
+        orography_data = np.where(land_sea_mask, orography_data, np.nan)
+
+        # Extract data array dimensions to define output arrays.
+        dataarray_shape = next(temperature_cube.slices([y_coord,
+                                                        x_coord])).shape
+        dataarray_size = dataarray_shape[0] * dataarray_shape[1]
+
+        # Array containing all of the subsections extracted from data array.
+        all_temp_subsections = np.zeros((dataarray_size,
+                                         self.footprintarray_size))
+        all_orog_subsections = np.zeros((dataarray_size,
+                                         self.footprintarray_size))
+
+        # Enforces single precision to speed up calculations.
+        all_temp_subsections = all_temp_subsections.astype(np.float32)
+        all_orog_subsections = all_orog_subsections.astype(np.float32)
+
+        # Check if a dimensional realization coordinate exists. If so, the
+        # cube is sliced, so that it becomes a scalar coordinate.
+        try:
+            _ = temperature_cube.coord('realization', dim_coords=True)
+        except iris.exceptions.CoordinateNotFoundError:
+            slices_over_realization = [temperature_cube]
+        else:
+            slices_over_realization = temperature_cube.slices_over(
+                "realization")
+
+        # Creates cube list to hold lapse rate data.
+        lapse_rate_cube_list = iris.cube.CubeList([])
+
+        for temp_slice in slices_over_realization:
+
+            # Create slice to store lapse rate values.
+            lapse_rate_slice = temp_slice
+
+            temperature_data = temp_slice.data
+
+            # Fill sea points with NaN values. Can't use Numpy mask since not
+            # recognised by "generic_filter" function.
+            temperature_data = np.where(land_sea_mask, temperature_data,
+                                        np.nan)
+
+            # cval is value given to points outside the array.
+            fnc = GenericFunc(allbuffers=all_temp_subsections)
+            generic_filter(temperature_data, fnc.filter,
+                           size=self.footprint_len,
+                           mode='constant', cval=np.nan)
```

Comment:
Again: we really need a more descriptive name for `GenericFunc`.  It's not at all clear what it's doing.